### PR TITLE
fix: open existing workspace when factory URL contains a /tree/<branc…

### PR DESCRIPTION
**Pull request by olexii4**
Fixes duplicate workspace creation when a factory URL contains a GitHub `/tree/<branch>` path and "Create New" is disabled.

### Root Cause

`getSameRepoWorkspaces()` in `CheckExistingWorkspaces` determines whether an existing workspace was created from the same repository by testing two conditions:

```typescript
workspace.source === factoryParams.sourceUrl && revision === factoryParams.revision
```

When a workspace is created from `https://github.com/org/repo/tree/my-branch`:

- `workspace.source` is stored as the full URL including `/tree/my-branch`.
- The devfile project is created with `checkoutFrom.revision = 'my-branch'` (extracted by `getProjectFromLocation.ts` which splits on `/tree/`).

When the user returns with the same URL and "Create New" unchecked:

- `factoryParams.sourceUrl` correctly equals `workspace.source` — the full URL including `/tree/my-branch`.
- `factoryParams.revision` is `undefined` because `getRevision()` only reads the explicit `?revision=` query parameter, never the `/tree/<branch>` path segment.

The revision mismatch (`'my-branch' !== undefined`) makes every existing workspace fail the filter, so `getSameRepoWorkspaces()` returns an empty list and a duplicate is created.

URLs without an explicit branch are unaffected because both `workspace.checkoutFrom.revision` and `factoryParams.revision` are `undefined` in that case.

### Fix

Extended `getRevision()` in `buildFactoryParams.ts` to also extract the branch name from the factory URL's `/tree/<branch>` path segment when no explicit `?revision=` query parameter is present:

```diff
 function getRevision(searchParams: URLSearchParams): string | undefined {
-  return searchParams.get(REVISION_ATTR) || undefined;
+  const revisionParam = searchParams.get(REVISION_ATTR) || undefined;
+  if (revisionParam) {
+    return revisionParam;
+  }
+  const factoryUrl = getFactoryUrl(searchParams);
+  if (factoryUrl) {
+    try {
+      const url = new URL(factoryUrl);
+      const parts = url.pathname.split('/tree/');
+      if (parts.length > 1 && parts[1]) {
+        return parts[1];
+      }
+    } catch { /* not a valid URL */ }
+  }
+  return undefined;
 }
```

The explicit `?revision=` query parameter still takes precedence.

### What issues does this PR fix or reference?

fixes https://redhat.atlassian.net/browse/CRW-10950

### Is it tested? How?

1. Deploy Dev Spaces with the dashboard image built from this PR.
2. Go to the User Dashboard and confirm **Create New** is enabled (default).
3. In **Create Workspace**, paste a GitHub URL that includes a branch path, e.g.:
   `https://github.com/<org>/<repo>/tree/<branch>`
   Create the workspace and wait for it to reach Running.
4. Return to **Create Workspace**.
5. Disable **Create New**.
6. Paste the same GitHub URL (same branch) and open the workspace.

Expected: the existing workspace is opened — no duplicate is created.

Before this fix: a new workspace would be created because the branch embedded in the
`/tree/<branch>` path was not matched against the existing workspace's revision, so
`getSameRepoWorkspaces()` always returned an empty list.

### CODE REVIEW

**Comment by tolusha** [comment]
Hi! I'm **che-ai-assistant** — I help with your pull requests.

**Available commands**:
- `/che-ai-assistant generate-che-doc` — Generate a documentation PR based on this PR's changes
- `/che-ai-assistant ok-pr-review` — Run a comprehensive PR review (summary, code review, deep review, impact analysis)
- `/che-ai-assistant help` — Show this help message

**Comment by olexii4** [comment]
/che-ai-assistant ok-pr-review

Review is complete. Please check the review comments below.

**Review by tolusha — APPROVED** [review]
## Comprehensive PR Review - All Stages Complete

✅ **APPROVED** - This is an excellent fix with clear root cause analysis, minimal and well-scoped changes, and comprehensive test coverage.

### Review Summary

**Standard Review**: ✅ Approve
**Deep Review**: ✅ Design is Sound  
**Impact Review**: ✅ No Concerns

### Highlights

- **Exemplary PR description** - Thorough root cause analysis with concrete examples of the workspace matching failure
- **Correct abstraction layer** - Fix placed in `getRevision()` where it belongs, maintaining separation of concerns
- **Follows existing patterns** - Uses the same `split('/tree/')` approach as `getProjectFromLocation.ts` for consistency
- **Strong test coverage** - 4 unit tests covering all scenarios (no param, query param, /tree/ path, precedence) plus integration test
- **Zero operational impact** - Pure correctness improvement with no new dependencies, no config changes, no API surface changes

### Optional Suggestions

The review identified a few optional improvements that would strengthen the test suite and documentation:

1. **Test for branch names with slashes** - Consider adding a test for `feature/branch-name` patterns (common Git convention, used in this very project)
2. **Test for trailing-slash edge case** - Document the behavior for malformed URLs like `/tree/` (correctly returns `undefined`)
3. **Document the shared pattern** - Add a cross-reference comment noting that both `getRevision()` and `getProjectFromLocation.ts` use `split('/tree/')`, so both need updating if this pattern changes

These are suggestions for future work, not blockers.

### System-Level Analysis

- **Observability**: No new logging/metrics needed (synchronous parsing function)
- **Failure handling**: Correct - try/catch returns `undefined` on parse failure, which is the right fallback
- **Resource cleanup**: N/A - no resources opened
- **Security**: No concerns - read-only URL parsing, no injection vectors
- **Performance**: Negligible - executes once during factory param construction

---

**Verdict**: Ready to merge. The fix correctly solves the duplicate workspace issue with no operational risk.

🤖 *Automated review by [che-ai-assistant](https://github.com/che-incubator/che-ai-assistant) powered by Claude Sonnet 4.5*

File: packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts [hunk 1]
```diff
@@ -173,7 +173,26 @@ function getRemotes(searchParams: URLSearchParams): string | undefined {
...
+  if (revisionParam) {
+    return revisionParam;
+  }
+  // Extract branch from a GitHub-style /tree/<branch> factory URL path so that
+  // getSameRepoWorkspaces() can match workspaces whose devfile projects were created
+  // with checkoutFrom.revision set to that branch name.
+  const factoryUrl = getFactoryUrl(searchParams);
+  if (factoryUrl) {
+    try {
+      const url = new URL(factoryUrl);
```
Comment (tolusha) [inline packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts:188]: The `split('/tree/')` approach correctly handles branch names containing slashes (e.g., `feature/my-branch`) by returning the full remainder. However, if a URL points to a specific file like `https://github.com/org/repo/tree/main/src/file.ts`, the extracted revision would be `main/src/file.ts` instead of `main`.

This matches the existing behavior in `getProjectFromLocation.ts` (consistent within the codebase), and GitHub disambiguates this server-side, so it's not a blocker. Worth noting for future work if supporting file-level URLs becomes a requirement.

File: packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts [hunk 2]
```diff
@@ -173,7 +173,26 @@ function getRemotes(searchParams: URLSearchParams): string | undefined {
 }
 
 function getRevision(searchParams: URLSearchParams): string | undefined {
-  return searchParams.get(REVISION_ATTR) || undefined;
+  const revisionParam = searchParams.get(REVISION_ATTR) || undefined;
+  if (revisionParam) {
+    return revisionParam;
+  }
+  // Extract branch from a GitHub-style /tree/<branch> factory URL path so that
```
Comment (tolusha) [inline packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts:180]: Consider adding a cross-reference to `getProjectFromLocation.ts` in this comment. Both locations use the same `split('/tree/')` pattern, so if this ever needs to change (e.g., to support Bitbucket's `/src/` paths), both need updating.

```suggestion
  // Extract branch from a GitHub-style /tree/<branch> factory URL path so that
  // getSameRepoWorkspaces() can match workspaces whose devfile projects were created
  // with checkoutFrom.revision set to that branch name.
  // Note: getProjectFromLocation.ts uses the same pattern for consistency.
```

This is optional - just helps future maintainability.

File: packages/dashboard-frontend/src/services/helpers/factoryFlow/__tests__/buildFactoryParams.spec.ts [hunk 3]
```diff
@@ -123,6 +125,38 @@ describe('buildFactoryParams', () => {
...
+        expect(buildFactoryParams(searchParams).revision).toBe('my-feature-branch');
+      });
+
+      it('should prefer ?revision= over the /tree/ branch in the factory URL', () => {
+        const searchParams = new URLSearchParams({
+          [FACTORY_URL_ATTR]: 'https://github.com/eclipse-che/che-dashboard/tree/url-branch',
+          [REVISION_ATTR]: 'param-branch',
+        });
+        expect(buildFactoryParams(searchParams).revision).toBe('param-branch');
+      });
```
Comment (tolusha) [inline packages/dashboard-frontend/src/services/helpers/factoryFlow/__tests__/buildFactoryParams.spec.ts:157]: Great test coverage! Two optional additions that would strengthen the test suite:

**1. Test for branch names with slashes** (common Git convention):
```typescript
it('should extract branch names containing slashes from /tree/ path', () => {
  const searchParams = new URLSearchParams({
    [FACTORY_URL_ATTR]: 'https://github.com/eclipse-che/che-dashboard/tree/feature/CRW-10950',
  });
  expect(buildFactoryParams(searchParams).revision).toBe('feature/CRW-10950');
});
```

**2. Test for trailing-slash edge case**:
```typescript
it('should return undefined for /tree/ with no branch name', () => {
  const searchParams = new URLSearchParams({
    [FACTORY_URL_ATTR]: 'https://github.com/eclipse-che/che-dashboard/tree/',
  });
  expect(buildFactoryParams(searchParams).revision).toBeUndefined();
});
```

Both cases are already handled correctly by the code - these tests would just document the expected behavior.

Comment (olexii4) [reply packages/dashboard-frontend/src/services/helpers/factoryFlow/__tests__/buildFactoryParams.spec.ts:157]: I have fixed it

File: packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts (hunk 2 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts:180]: I have fixed it

File: packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts (hunk 1 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/services/helpers/factoryFlow/buildFactoryParams.ts:188]: I have fixed it

**Review by akurinnoy — APPROVED** [review]
